Web Automation: generate WebDriver BiDi protocol backend dispatchers from JSON input files

burg opened the pull request:
#### 5f0f237529e13e34275420ae27e1a3e029d56c7f
<pre>
Web Automation: generate WebDriver BiDi protocol backend dispatchers from JSON input files
<a href="https://bugs.webkit.org/show_bug.cgi?id=284322">https://bugs.webkit.org/show_bug.cgi?id=284322</a>
&lt;<a href="https://rdar.apple.com/136766847">rdar://136766847</a>&gt;

Reviewed by Devin Rousso and Patrick Angle.

This change introduces two domain specifications for WebDriver BiDi and hooks up
the protocol code generator to produce C++ backend dispatchers for each domain.
Previously, Automation.json was used to produce an Automation protocol dispatcher.
To support WebDriver BiDi&apos;s concept of multiple domains, we take an approach similar
to Web Inspector: combine all the domain definitions into one big JSON file and
call the code generator. Introduce a new class called WebDriverBidiProcessor which
dispatches WebDriver BiDi messages to the available backend domain dispatchers.

In order to interoperate within the existing Automation protocol, we take the nested approach
of adding a command and event to Automation protocol for WebDriverBidi message send/receive.
On the driver side, delivery of the message from Automation.bidiMessageSent to the client is further
mediated by an event publish/subscribe mechanism that matches expected semantics of WebDriver BiDi.

We do not try to handle event pub/sub from UIProcess, as these subscribe/unsubscribe requests can
come in from a session client prior to a browser being launched. Secondly, pub/sub state is per-client
connection, rather than per-session, so this cannot really be mediated from UIProcess.

* Source/WTF/wtf/PlatformEnableCocoa.h: Define ENABLE_WEBDRIVER_BIDI for Cocoa ports.

* Source/WebKit/DerivedSources.make:
For the most part, we can copy the approach taken to generate Inspector Protocol files.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
(CppBackendDispatcherHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_protocol_types_header.py:
(CppProtocolTypesHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Framework.fromString):
(Frameworks):
* Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
(generate_from_specification):
Add a new target pseudo-framework called &apos;WebDriverBidi&apos; which will be used to generate WebDriver BiDi
protocol code.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
Add new files.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
Added, let&apos;s start with Browser.close and BrowsingContext.navigate to flesh things out.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::WebAutomationSession): Create a Bidi dispatcher.
(WebKit::WebAutomationSession::processBidiMessage): Added.
(WebKit::WebAutomationSession::sendBidiMessage): Added.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp: Added.
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::~WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::protectedFrontendRouter const):
(WebKit::WebDriverBidiProcessor::protectedBackendDispatcher const):
(WebKit::WebDriverBidiProcessor::processBidiMessage):
(WebKit::WebDriverBidiProcessor::sendBidiMessage):
(WebKit::WebDriverBidiProcessor::sendMessageToFrontend):
(WebKit::WebDriverBidiProcessor::navigate):
(WebKit::WebDriverBidiProcessor::close):

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
Copy pre-existing SaferCPP expectations for generated inspector protocol code.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add new Derived Sources files.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Add existing AutomationInstrumentation
files to the Xcode project since we need to start building with ENABLE(WEBDRIVER_BIDI) on Cocoa.

Web Automation: start using WebDriver BiDi generated frontend dispatchers from JSON input files
<a href="https://bugs.webkit.org/show_bug.cgi?id=284322">https://bugs.webkit.org/show_bug.cgi?id=284322</a>
&lt;<a href="https://rdar.apple.com/136788420">rdar://136788420</a>&gt;

Reviewed by Devin Rousso and Patrick Angle.

This change starts using generated frontend dispatchers as the mechanism to deliver per-domain
WebDriver BiDi events from WebDriverBidiProcessor out through Automation protocol.

Hook up the recently implemented log.entryAdded event so that this event also is sent out via
the newly added WebDriverBidi protocol (`Automation.bidiMessageSent`). When clients have moved
away from using `Automation.logEntryAdded`, that event can be removed in favor of the event that
is now defined in `BidiLog.json`.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_header.py:
(CppFrontendDispatcherHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_implementation.py:
(CppFrontendDispatcherImplementationGenerator._generate_secondary_header_includes):
Mark these headers as allowed to be included for WebDriverBidi. (Otherwise, it won&apos;t build as-is.)
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make: Add new domain.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::logEntryAdded):
Additionally send the log entry to be dispatched via the generated BidiLogFrontendDispatcher
interface. It will eventually come back out via `Automation.bidiMessageSent`.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiLog.json: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/290582@main">https://commits.webkit.org/290582@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/868716f48dc58cff1176b56365b684e0c262692e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90515 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95526 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41297 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18366 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69654 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27214 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50002 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40427 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83321 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37530 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97355 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89295 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17707 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17965 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77915 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77869 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19233 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22311 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17717 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111784 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17456 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20911 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19240 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

the-chenergy commented on Source/JavaScriptCore/inspector/scripts/codegen/models.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 #
-# Copyright (c) 2014 Apple Inc. All rights reserved.
+# Copyright (c) 2014-2024 Apple Inc. All rights reserved.
```
2025?

AnujPanta1 commented on Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 #
-# Copyright (c) 2014, 2016 Apple Inc. All rights reserved.
+# Copyright (c) 2014-2024 Apple Inc. All rights reserved.
```
same nit as Qianlang if it does become 2025 now

dcrousso commented on Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
```diff
@@ -76,12 +76,12 @@ def generate_output(self):
-            (["JavaScriptCore", "WebKit"], ("WTF", "wtf/text/MakeString.h")),
-            (["JavaScriptCore", "WebKit"], ("WTF", "wtf/text/WTFString.h")),
-            (["JavaScriptCore", "WebKit"], ("std", "tuple")),
+            (["JavaScriptCore", "WebKit", "WebKitBidi"], (self.model().framework.name, "%sProtocolObjects.h" % self.protocol_name())),
```
NIT: can we call this `WebDriverBidi` instead?  i think it'd be more clear for someone without context reading this in the future

dcrousso commented on Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
```diff
@@ -184,6 +184,14 @@ def load_specification(protocol, filepath, isSupplemental=False):
         generators.append(ObjCProtocolTypeConversionsImplementationGenerator(*generator_arguments))
         generators.append(ObjCProtocolTypesImplementationGenerator(*generator_arguments))
 
+    elif protocol.framework is Frameworks.WebKitBidi and generate_backend:
```
since this is the same as `Frameworks.WebKit` can we merge these 2 `if`?

dcrousso commented on Source/JavaScriptCore/inspector/scripts/tests/expected/command-targetType-matching-domain-debuggableType.json-result:
```diff
@@ -36,6 +36,45 @@ InspectorBackend.registerDomainDispatcher = InspectorBackend.registerDispatcher.
 InspectorBackend.activateDomain("Domain", ["page"]);
 ### End File: InspectorBackendCommands.js.in
 
+### Begin File: InspectorBackendCommands.js.out
```
can we make the test rebase stuff a separate PR?

dcrousso commented on Source/WebKit/DerivedSources.make:
```diff
@@ -467,7 +467,7 @@ NOTIFICATION_ALLOW_LISTS = \
 	mv $@.tmp $@
 
-AUTOMATION_PROTOCOL_GENERATOR_SCRIPTS = \
+JSON_RPC_GENERATOR_SCRIPTS = \
```
i feel like there should be an `AUTOMATION` or `WEBDRIVER` or something in this name so as to help future readers without context understand what it's for

dcrousso replied:
so are we not gonna have `AUTOMATION` or `WEBDRIVER` in this name?

burg replied:
I'd rather not, as those are specific protocol names, which can be seen in context by following the target dependencies.

(It would be nice to move more of the codegen machinery out of Inspector namespace but that's a longer term project.)

dcrousso commented on Source/WebKit/DerivedSources.make:
```diff
@@ -510,6 +513,38 @@ all : $(AUTOMATION_PROTOCOL_OUTPUT_FILES)
 all : WebAutomationSessionProxyScriptSource.h
 
+# Much like Web Inspector protocol, WebDriver Bidi protocol consists of multiple domains. Each domain has one .json specification file.
+BIDI_PROTOCOL_INPUT_FILES = \
```
ditto with all of these (especially to prevent confusion with bidi text)

perhaps these could be `AUTOMATION_BIDI_*` or `WEBDRIVER_BIDI_*`?

burg replied:
WEBDRIVER_BIDI it is..

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
```diff
@@ -0,0 +1,12 @@
+        {
+            "name": "close",
+            "description": "Terminates the WebDriver session to which it is sent.",
+            "parameters": []
```
NIT: this can be omitted

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -0,0 +1,61 @@
+            "description": "Unique string identifying an ongoing navigation."
+        },
+        {
+          "id": "ReadinessState",
```
Style: indentation

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -0,0 +1,61 @@
+                {
+                    "name": "navigation",
+                    "$ref": "BrowsingContext.Navigation",
+                    "optional": true,
```
it's kinda odd to have optional things come before non-optional stuff, so id recommend flipping the order of this with `url`

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -0,0 +1,61 @@
+                    "description": "The URL the browsing context navigated to."
+                }
+            ],
+            "async": true
```
i thought you didnt like `async`?  or is it just how it's currently implemented that's not ideal?

burg replied:
The current implementation is not ideal.

dcrousso commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -985,6 +985,10 @@
 #define ENABLE_WEBDRIVER_WHEEL_INTERACTIONS 1
 #endif
 
+#if !defined(ENABLE_WEBDRIVER_BIDI) && (PLATFORM(MAC) || PLATFORM(IOS) || PLATFORM(IOS_SIMULATOR) || PLATFORM(VISION))
```
so everything except `MAC_CATALYST`?

burg replied:
no TV, watch, or catalyst

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -0,0 +1,61 @@
+                    "name": "context",
+                    "$ref": "BrowsingContext",
+                    "description": "The identifier of the browsing context to navigate."
+                },
```
Style: I much prefer having the values in `"parameters"` and `"returns"` be inlined instead of spread across multiple lines as it makes it easier to see parameters at a quick glance instead of having to scan so many more lines of code).  is there a reason for these being spread across multiple lines?

burg replied:
LOL, I completely forgot that we do inline these in Web Inspector's files.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -0,0 +1,146 @@
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
```
is there a point of even having this stub if it's not implemented?  i.e. would it be better to not have this here at all and instead wait for another PR that fully implements it?

burg replied:
Yeah, I had a hacky implementation in my working copy but I dropped it since it will be rewritten soon. The reason that this stub exists is so that I could validate command generation for multiple domains.

patrickangle commented on Source/WebKit/UIProcess/Automation/Automation.json:
```diff
@@ -937,6 +945,14 @@
             ]
+        },
+        {
+            "name": "bidiMessageSent",
```
The names `processBidiMessage` and `bidiMessageSent` names don't really feel like the call and response that they are. Could we consider `bidiCommand` and `bidiResponse`, which frames both of these from the client's point of view.

lauromoura replied:
What about `processBiDiCommand` instead of `processBiDiMessage`, as the methods usually start with a verb and all messages that come from the client are commands?

One drawback of `bidiResponse` is in the case of non-response messages, like events.

lauromoura commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -142,6 +146,9 @@ WebAutomationSession::WebAutomationSession()
     , m_backendDispatcher(BackendDispatcher::create(m_frontendRouter.copyRef()))
     , m_domainDispatcher(AutomationBackendDispatcher::create(m_backendDispatcher, this))
     , m_domainNotifier(makeUnique<AutomationFrontendDispatcher>(m_frontendRouter))
+#if ENABLE(WEBDRIVER_BIID)
```
```suggestion
#if ENABLE(WEBDRIVER_BIDI)
```

gsnedders commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -985,6 +985,10 @@
 #define ENABLE_WEBDRIVER_WHEEL_INTERACTIONS 1
 #endif
 
+#if !defined(ENABLE_WEBDRIVER_BIDI) && (PLATFORM(MAC) || PLATFORM(IOS) || PLATFORM(IOS_SIMULATOR) || PLATFORM(VISION))
```
`PLATFORM(IOS_SIMULATOR)` is redundant here, as `PLATFORM(IOS)` already includes the Simulator.

lauromoura commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -142,6 +146,9 @@ WebAutomationSession::WebAutomationSession()
     , m_domainDispatcher(AutomationBackendDispatcher::create(m_backendDispatcher, this))
     , m_domainNotifier(makeUnique<AutomationFrontendDispatcher>(m_frontendRouter))
+#if ENABLE(WEBDRIVER_BIID)
+    , m_bidiProcessor(makeUnique<WebDriverBidiProcessor>(this))
```
```suggestion
    , m_bidiProcessor(makeUnique<WebDriverBidiProcessor>(*this))
```

lauromoura reviewed (changes requested):
I'm not an expert on generated inspector-related code but looks good overall. Requesting changes due to two small issues found while testing locally as they flew under EWS's radar.

Btw, I'm moving the implementation I was doing for `BrowsingContext.getTree` to this scheme to experiment with it.

burg committed:
Web Automation: generate WebDriver BiDi protocol backend dispatchers from JSON input files
https://bugs.webkit.org/show_bug.cgi?id=284322
<rdar://136766847>

Reviewed by Devin Rousso and Patrick Angle.

This change introduces two domain specifications for WebDriver BiDi and hooks up
the protocol code generator to produce C++ backend dispatchers for each domain.
Previously, Automation.json was used to produce an Automation protocol dispatcher.
To support WebDriver BiDi's concept of multiple domains, we take an approach similar
to Web Inspector: combine all the domain definitions into one big JSON file and
call the code generator. Introduce a new class called WebDriverBidiProcessor which
dispatches WebDriver BiDi messages to the available backend domain dispatchers.

In order to interoperate within the existing Automation protocol, we take the nested approach
of adding a command and event to Automation protocol for WebDriverBidi message send/receive.
On the driver side, delivery of the message from Automation.bidiMessageSent to the client is further
mediated by an event publish/subscribe mechanism that matches expected semantics of WebDriver BiDi.

We do not try to handle event pub/sub from UIProcess, as these subscribe/unsubscribe requests can
come in from a session client prior to a browser being launched. Secondly, pub/sub state is per-client
connection, rather than per-session, so this cannot really be mediated from UIProcess.

* Source/WTF/wtf/PlatformEnableCocoa.h: Define ENABLE_WEBDRIVER_BIDI for Cocoa ports.

* Source/WebKit/DerivedSources.make:
For the most part, we can copy the approach taken to generate Inspector Protocol files.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
(CppBackendDispatcherHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_protocol_types_header.py:
(CppProtocolTypesHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Framework.fromString):
(Frameworks):
* Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
(generate_from_specification):
Add a new target pseudo-framework called 'WebDriverBidi' which will be used to generate WebDriver BiDi
protocol code.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
Add new files.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
Added, let's start with Browser.close and BrowsingContext.navigate to flesh things out.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::WebAutomationSession): Create a Bidi dispatcher.
(WebKit::WebAutomationSession::processBidiMessage): Added.
(WebKit::WebAutomationSession::sendBidiMessage): Added.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp: Added.
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::~WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::protectedFrontendRouter const):
(WebKit::WebDriverBidiProcessor::protectedBackendDispatcher const):
(WebKit::WebDriverBidiProcessor::processBidiMessage):
(WebKit::WebDriverBidiProcessor::sendBidiMessage):
(WebKit::WebDriverBidiProcessor::sendMessageToFrontend):
(WebKit::WebDriverBidiProcessor::navigate):
(WebKit::WebDriverBidiProcessor::close):

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
Copy pre-existing SaferCPP expectations for generated inspector protocol code.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add new Derived Sources files.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Add existing AutomationInstrumentation
files to the Xcode project since we need to start building with ENABLE(WEBDRIVER_BIDI) on Cocoa.

Web Automation: start using WebDriver BiDi generated frontend dispatchers from JSON input files
https://bugs.webkit.org/show_bug.cgi?id=284322
<rdar://136788420>

Reviewed by Devin Rousso and Patrick Angle.

This change starts using generated frontend dispatchers as the mechanism to deliver per-domain
WebDriver BiDi events from WebDriverBidiProcessor out through Automation protocol.

Hook up the recently implemented log.entryAdded event so that this event also is sent out via
the newly added WebDriverBidi protocol (`Automation.bidiMessageSent`). When clients have moved
away from using `Automation.logEntryAdded`, that event can be removed in favor of the event that
is now defined in `BidiLog.json`.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_header.py:
(CppFrontendDispatcherHeaderGenerator._generate_secondary_header_includes):
* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_implementation.py:
(CppFrontendDispatcherImplementationGenerator._generate_secondary_header_includes):
Mark these headers as allowed to be included for WebDriverBidi. (Otherwise, it won't build as-is.)
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make: Add new domain.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::logEntryAdded):
Additionally send the log entry to be dispatched via the generated BidiLogFrontendDispatcher
interface. It will eventually come back out via `Automation.bidiMessageSent`.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiLog.json: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/290582@main
